### PR TITLE
improved docker image publication

### PR DESCRIPTION
**Pull request by YannickWeber**

### CODE REVIEW

File: .github/workflows/snapshot.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+name: Publish Snapshot to Docker Hub
+
+on:
+  push:
+    branches:
+      - master
+
+concurrency:
+  group: ${{ github.ref }}-snapshot
+  cancel-in-progress: true
```
Comment (SgtSilvio) [inline .github/workflows/snapshot.yml:10]: is this really necessary? how could this be triggered multiple times for the same ref?

File: .github/workflows/snapshot.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+jobs:
+  snapshot:
+    environment: mavenCentralPublish
```
Comment (SgtSilvio) [inline .github/workflows/snapshot.yml:14]:
```suggestion
```
should not be necessary here

File: docker/build.sh [hunk 3]
```diff
@@ -4,11 +4,6 @@ set -eo pipefail
 
 IMAGE=${TARGET_IMAGE:-hivemq/hivemq-ce:snapshot}
```
Comment (A-Imal) [inline docker/build.sh:5]: This whole script file is not used anymore.

File: docker/publish-snapshot.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
+
+set -eo pipefail
+
+IMAGE=${TARGET_IMAGE:-hivemq/hivemq-ce:snapshot}
```
Comment (A-Imal) [inline docker/publish-snapshot.sh:5]:
```suggestion
IMAGE="-hivemq/hivemq-ce:snapshot"
```

File: docker/publish-snapshot.sh [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+cd docker
+cp ../build/zip/hivemq-ce-${HIVEMQ_VERSION}.zip .
+docker build --build-arg HIVEMQ_VERSION=${HIVEMQ_VERSION} -f Dockerfile -t ${IMAGE} .
+rm -f hivemq-ce-${HIVEMQ_VERSION}.zip
+
+echo "Tagging image as snapshot"
+TAGGED_IMAGE="${IMAGE//:*}:snapshot"
+docker tag ${IMAGE} "${TAGGED_IMAGE}"
+
+
```
Comment (A-Imal) [inline docker/publish-snapshot.sh:20]:
```suggestion
```

File: docker/publish.sh [hunk 6]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+
+set -eo pipefail
+
+IMAGE=${TARGET_IMAGE:-hivemq/hivemq-ce:}
```
Comment (A-Imal) [inline docker/publish.sh:5]:
```suggestion
IMAGE_PREFIX="hivemq/hivemq-ce"
LATEST_IMAGE="${IMAGE_PREFIX}:latest"
```

File: docker/publish.sh [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+cd "$(dirname $0)/../"
+HIVEMQ_VERSION=$(./gradlew properties | grep ^version: | sed -e "s/version: //")
+echo "Building Docker image for HiveMQ ${HIVEMQ_VERSION}"
+./gradlew hivemqZip
+cd docker
+cp ../build/zip/hivemq-ce-${HIVEMQ_VERSION}.zip .
+docker build --build-arg HIVEMQ_VERSION=${HIVEMQ_VERSION} -f Dockerfile -t ${IMAGE} .
```
Comment (A-Imal) [inline docker/publish.sh:13]:
```suggestion
docker build --build-arg HIVEMQ_VERSION=${HIVEMQ_VERSION} -t ${LATEST_IMAGE} .
```

File: docker/publish.sh [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+rm -f hivemq-ce-${HIVEMQ_VERSION}.zip
+
+echo "Tagging image as ${HIVEMQ_VERSION}"
+TAGGED_IMAGE="${IMAGE//:*}:${HIVEMQ_VERSION}"
```
Comment (A-Imal) [inline docker/publish.sh:17]:
```suggestion
VERSIONED_IMAGE="${IMAGE_PREFIX}:${HIVEMQ_VERSION}"
```

File: docker/publish.sh [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+docker tag ${IMAGE} "${TAGGED_IMAGE}"
```
Comment (A-Imal) [inline docker/publish.sh:18]:
```suggestion
docker tag ${LATEST_IMAGE} ${VERSIONED_IMAGE}
```

File: docker/publish.sh [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    echo "Pushing image as ${TAGGED_IMAGE}"
+    docker push ${TAGGED_IMAGE}
+fi
+
+# If we're building a tagged commit, it will also be the new :latest image on Docker Hub.
+if [[ ${PUSH_IMAGE} == true ]]; then
+    echo "Tagging image as ${IMAGE//:*}:latest"
+    docker tag ${IMAGE} "${IMAGE//:*}:latest"
+    echo "Pushing image as ${IMAGE//:*}:latest"
+    docker push ${IMAGE//:*}:latest
```
Comment (A-Imal) [inline docker/publish.sh:30]:
```suggestion
    echo "Pushing image as ${VERSIONED_IMAGE}"
    docker push ${VERSIONED_IMAGE}
    
    echo "Pushing image as ${LATEST_IMAGE}"
    docker push ${LATEST_IMAGE}
```

**Review by A-Imal — CHANGES_REQUESTED** [review]

File: docker/publish.sh [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+fi
```
Comment (A-Imal) [inline docker/publish.sh:31]:
```suggestion
if [[ ${PUSH_IMAGE} == true ]]; then
    echo "Pushing image as ${VERSIONED_IMAGE}"
    docker push ${VERSIONED_IMAGE}
    
    echo "Pushing image as ${LATEST_IMAGE}"
    docker push ${LATEST_IMAGE}
fi
```

**Review by A-Imal — CHANGES_REQUESTED** [review]

File: docker/build.sh (hunk 3 above)
Comment (YannickWeber) [reply docker/build.sh]: It is documented in the readme, that you can use this script to create a docker image locally.

File: docker/publish-snapshot.sh [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+IMAGE="hivemq/hivemq-ce:snapshot"
+
+cd "$(dirname $0)/../"
+HIVEMQ_VERSION=$(./gradlew properties | grep ^version: | sed -e "s/version: //")
+echo "Building Docker image for HiveMQ ${HIVEMQ_VERSION}"
+./gradlew hivemqZip
+cd docker
+cp ../build/zip/hivemq-ce-${HIVEMQ_VERSION}.zip .
+docker build --build-arg HIVEMQ_VERSION=${HIVEMQ_VERSION} -f Dockerfile -t ${IMAGE} .
+rm -f hivemq-ce-${HIVEMQ_VERSION}.zip
```
Comment (A-Imal) [inline docker/publish-snapshot.sh:14]: Can be removed as the workspace is ephemeral.

File: docker/publish.sh [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+LATEST_IMAGE="${IMAGE_PREFIX}:latest"
+
+cd "$(dirname $0)/../"
+HIVEMQ_VERSION=$(./gradlew properties | grep ^version: | sed -e "s/version: //")
+echo "Building Docker image for HiveMQ ${HIVEMQ_VERSION}"
+./gradlew hivemqZip
+cd docker
+cp ../build/zip/hivemq-ce-${HIVEMQ_VERSION}.zip .
+docker build --build-arg HIVEMQ_VERSION=${HIVEMQ_VERSION} -f Dockerfile -t ${LATEST_IMAGE} .
+rm -f hivemq-ce-${HIVEMQ_VERSION}.zip
```
Comment (A-Imal) [inline docker/publish.sh:15]: Can be removed as the workspace is ephemeral.

**Review by A-Imal — APPROVED** [review]

File: docker/publish-snapshot.sh [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+
+if [[ ${PUSH_IMAGE} == true ]]; then
+    echo "Pushing image"
+    docker push ${IMAGE}
+fi
```
Comment (SgtSilvio) [inline docker/publish-snapshot.sh:19]: This file is now basically the same as build.sh + the additional if statement, we could just move the if to build.sh to remove duplication

**Review by A-Imal — APPROVED** [review]

**Review by SgtSilvio — APPROVED** [review]

**Pull request merged**